💄 Add option for panels to fit to content

harkabeeparolus opened the pull request:
Provide an option for using Rich `Panel.fit()` aka `Panel(expand=False)` for help panels and errors. This was mentioned in discussion #607 and #680. To opt into this new, non-default behavior, just use:

```python
app = typer.Typer(rich_expand=False)
```

I think this can look nicer on very wide terminals, especially for CLIs with few parameters, or short parameter descriptions, so I would prefer to have the option to choose this style sometimes.

### Status

Ready to merge.

* The code worked fine on all the examples I tried from the Typer tutorial
* Docs and tests are finished, thanks to svlandeg!

### Example

![rich_expand_example](https://github.com/user-attachments/assets/8fd8b295-0b58-4b54-baf2-9a88dfc3bc85)

harkabeeparolus committed:
Add rich_expand parameter

harkabeeparolus committed:
Merge branch 'fastapi:master' into feature/rich_expand

harkabeeparolus committed:
Add documenation

harkabeeparolus committed:
Merge branch 'feature/rich_expand' of https://github.com/harkabeeparolus/typer into feature/rich_expand

harkabeeparolus committed:
Merge branch 'master' into feature/rich_expand

mshonichev commented:
Any chances to merge this PR soon? Can someone please restart CI?

svlandeg committed:
Merge branch 'master' into feature/rich_expand

harkabeeparolus commented:
> Any chances to merge this PR soon?

I believe what's missing is test coverage for my new code and logic, which is not my strong suit. I would appreciate any help or suggestions on how to approach writing tests for this.

harkabeeparolus committed:
Merge branch 'master' into feature/rich_expand

harkabeeparolus committed:
Fix AssertionError: Sum of ratios must be > 0

harkabeeparolus committed:
Add rich_expand to typer.run()

svlandeg committed:
Merge branch 'master' into feature/rich_expand

svlandeg commented:
> the only failing CI tests right now are on 3.8 and 3.9, which are end of life. Should we address these?

Yes, of course, as long as the CI is running 3.8/3.9 tests, that means we're still officially supporting these versions of Python, even if they've become EOL. We'll likely drop them at some point in the near future, but nevertheless every PR should always pass the full test suite.

svlandeg committed:
Use Union notation to avoid failing tests

svlandeg commented on typer/rich_utils.py:
```diff
@@ -234,7 +234,8 @@ def _get_parameter_help(
     markup_mode: MarkupMode,
-) -> Columns:
+    rich_expand: bool = True,
+) -> Union[Table, Column]:
```
```suggestion
) -> Union[Table, Columns]:
```

svlandeg committed:
Fix typo

harkabeeparolus commented on typer/rich_utils.py:
```diff
@@ -457,7 +467,7 @@ class MetavarHighlighter(RegexHighlighter):
             highlight=True,
             show_header=False,
-            expand=True,
+            expand=False,
```
That's a very good question. I'm not sure! I wrote that change almost a year ago, but honestly, It seems to have no effect on the result, after the last bugfix I made. Let me test it some more with the current code. That change should probably be removed!

harkabeeparolus committed:
Revert "Add rich_expand to typer.run()"

This reverts commit 7f9e5a407071204bd852b981ae547ce0cde1b47a.

harkabeeparolus committed:
Remove unneccessary defalt parameter values

harkabeeparolus committed:
Remove unnecessary default value

Co-authored-by: Sofie Van Landeghem <svlandeg@users.noreply.github.com>

harkabeeparolus committed:
Restore unneeded expand=False in options_table

svlandeg committed:
Merge branch 'master' into feature/rich_expand

svlandeg committed:
move new section to the end of the page

svlandeg committed:
add tutorial009 example

svlandeg committed:
Merge branch 'master' into feature/rich_expand

svlandeg committed:
update emoji

svlandeg committed:
show rich_expand=False results

svlandeg committed:
update example (simplify)

svlandeg committed:
revert

svlandeg committed:
fix test

svlandeg committed:
Merge branch 'feature/rich_expand' of https://github.com/harkabeeparolus/typer into feature/rich_expand

svlandeg committed:
fix outline

svlandeg reviewed (approved):
This now looks good to me, though another review from someone who hasn't worked on it directly would be good 🙏 

(Also @harkabeeparolus: feel free to check and report whether the docs look good to you after my edits!)

Current docs link can be found [here](https://github.com/fastapi/typer/pull/1096#issuecomment-3520528501).

harkabeeparolus commented:
Yeah, the docs look great to me! The example really helps to illustrate the different look.

@svlandeg, thank you so much! You've been super helpful and I'm really happy. 😊

svlandeg committed:
Merge branch 'master' into feature/rich_expand

svlandeg committed:
use new py39 format for tutorial file

harkabeeparolus committed:
Merge branch 'master' into feature/rich_expand

harkabeeparolus committed:
Merge remote-tracking branch 'upstream/master' into feature/rich_expand

pre-commit-ci-lite[bot] committed:
🎨 Auto format

harkabeeparolus committed:
🎨 Upgrade rich_expand typing to Python 3.10+

Rename tutorial009_py39 → tutorial009_py310 and update Union[] annotations
to X | Y syntax, matching upstream's migration in #1547/#1549.

harkabeeparolus committed:
Merge branch 'master' into feature/rich_expand

harkabeeparolus committed:
Merge branch 'master' into feature/rich_expand

harkabeeparolus committed:
Merge branch 'master' into feature/rich_expand
